### PR TITLE
Metrics Collector: Support for configurable transport type

### PR DESCRIPTION
**Pull request by and-rewsmith**
This PR:
1. Makes transport type configurable
2. Fixes bug where metrics collector will restart if cannot connect to EdgeHub when using AzureMonitor upload target. Now it will log a warning and continue in case of failed connection to EdgeHub. This connection is only used for internal diagnostics so it can be treated as recoverable.

Manually tested:
- Case where can switch between transport types
- Case where can work when opposite protocol head disabled in EdgeHub
- Case where can work when used protocol head disabled in EdgeHub and upload target is log analytics
- Case where won't work when used protocol head disabled in EdgeHub and upload target is iot message

## Azure IoT Edge PR checklist:

This checklist is used to make sure that common guidelines for a pull request are followed.

### General Guidelines and Best Practices
- [x] I have read the [contribution guidelines](https://github.com/azure/iotedge#contributing).
- [x] Title of the pull request is clear and informative.
- [x] Description of the pull request includes a concise summary of the enhancement or bug fix.

### Testing Guidelines
- [x] Pull request includes test coverage for the included changes.
- Description of the pull request includes 
	- [x] concise summary of tests added/modified
	- [x] local testing done.  

### Draft PRs
- Open the PR in `Draft` mode if it is:
	- Work in progress or not intended to be merged.
	- Encountering multiple pipeline failures and working on fixes.

_Note: We use the kodiakhq bot to merge PRs once the necessary checks and approvals are in place. When it merges a PR, kodiakhq converts the PR title to the commit title, PR description to the commit description, and squashes all the commits in the PR to a single commit. The net effect is that entire PR becomes a single commit. Please follow the best practices mentioned [here](https://chris.beams.io/posts/git-commit/#:~:text=The%20seven%20rules%20of%20a%20great%20Git%20commit,what%20and%20why%20vs.%20how%20For%20example%3A%20) for the PR title and description_

### CODE REVIEW

File: edge-modules/metrics-collector/src/ModuleClientWrapper.cs [hunk 1]
```diff
@@ -74,5 +68,35 @@ public void Dispose()
...
+                    case TransportType.Amqp_WebSocket_Only:
+                        return new ITransportSettings[] { new AmqpTransportSettings(TransportType.Amqp_WebSocket_Only) };
+                    default:
+                        return new ITransportSettings[] { new AmqpTransportSettings(TransportType.Amqp_Tcp_Only) };
+                }
+            }
+
+            ITransportSettings[] settings = GetTransportSettings();
+            ModuleClient moduleClient = await ModuleClient.CreateFromEnvironmentAsync(settings);
+            moduleClient.ProductInfo = Constants.ProductInfo;
```
Comment (and-rewsmith) [inline edge-modules/metrics-collector/src/ModuleClientWrapper.cs:94]: Product info is set correctly (i.e. before client is open async)

File: edge-modules/metrics-collector/src/ModuleClientWrapper.cs [hunk 2]
```diff
@@ -74,5 +68,35 @@ public void Dispose()
...
+            await moduleClient.OpenAsync();
+
+            LoggerUtil.Writer.LogInformation($"Successfully initialized module client of transport type [{transportType}]");
```
Comment (damonbarry) [inline edge-modules/metrics-collector/src/ModuleClientWrapper.cs:97]:
```suggestion
            LoggerUtil.Writer.LogInformation($"Successfully initialized module client using transport type [{transportType}]");
```

File: edge-modules/metrics-collector/src/Program.cs [hunk 3]
```diff
@@ -45,15 +45,27 @@ static async Task<int> MainAsync()
...
-
-                PeriodicTask periodicIothubConnect = new PeriodicTask(moduleClientWrapper.RecreateClientAsync, Settings.Current.IotHubConnectFrequency, TimeSpan.FromMinutes(1), LoggerUtil.Writer, "Reconnect to IoT Hub", true);
+                try
+                {
+                    moduleClientWrapper = await ModuleClientWrapper.BuildModuleClientWrapperAsync(Settings.Current.TransportType);
+                    PeriodicTask periodicIothubConnect = new PeriodicTask(moduleClientWrapper.RecreateClientAsync, Settings.Current.IotHubConnectFrequency, TimeSpan.FromMinutes(1), LoggerUtil.Writer, "Reconnect to IoT Hub", true);
+                }
+                catch (Exception e)
+                {
+                    String msg = String.Format("Error connecting to EdgeHub. Is EdgeHub up and running with settings for clients connecting over {0}? Exception: {1}", Settings.Current.TransportType, e);
```
Comment (damonbarry) [inline edge-modules/metrics-collector/src/Program.cs:58]:
```suggestion
                    String msg = String.Format("Error connecting to Edge Hub. Is Edge Hub up and running with settings for clients connecting over {0}? Exception: {1}", Settings.Current.TransportType, e);
```

File: edge-modules/metrics-collector/src/Program.cs [hunk 4]
```diff
@@ -45,15 +45,27 @@ static async Task<int> MainAsync()
...
             {
-                moduleClientWrapper = await ModuleClientWrapper.BuildModuleClientWrapperAsync(transportSettings);
-
-                PeriodicTask periodicIothubConnect = new PeriodicTask(moduleClientWrapper.RecreateClientAsync, Settings.Current.IotHubConnectFrequency, TimeSpan.FromMinutes(1), LoggerUtil.Writer, "Reconnect to IoT Hub", true);
+                try
+                {
+                    moduleClientWrapper = await ModuleClientWrapper.BuildModuleClientWrapperAsync(Settings.Current.TransportType);
+                    PeriodicTask periodicIothubConnect = new PeriodicTask(moduleClientWrapper.RecreateClientAsync, Settings.Current.IotHubConnectFrequency, TimeSpan.FromMinutes(1), LoggerUtil.Writer, "Reconnect to IoT Hub", true);
+                }
+                catch (Exception e)
```
Comment (damonbarry) [inline edge-modules/metrics-collector/src/Program.cs:56]: Are there exceptions that _aren't_ related to connecting to Edge Hub, that should cause the module to exit here, even if (UploadTarget == IotMessage)?

Comment (and-rewsmith) [reply edge-modules/metrics-collector/src/Program.cs:56]: I assume you mean "even if (UploadTarget == AzureMonitor)". For upload target of IotMessage we will always fail fast if a problem in this try block.

My logic was that whatever went wrong had to be either the module client creation or the periodic task creation, both of which apply only for the client which isn't necessary to be used for AzureMonitor upload path. So I think that there won't be any reason to exit and fail fast.

Comment (damonbarry) [reply edge-modules/metrics-collector/src/Program.cs:56]: > I assume you mean "even if (UploadTarget == AzureMonitor)"

Yes. 😏

> My logic was

Got it...

**Review by damonbarry — APPROVED** [review]

**Pull request merged**